### PR TITLE
Tweak SemaphoreDisposer to null out the held semaphore

### PR DESCRIPTION
**Pull request by jasonmalinowski**
While investigating a deadlock, it would have been made the analysis easier if we null out the semaphore once we've released it, because it means the then one-and-only instance of a SemaphoreDisposer that points to the semaphore is the one that is holding the lock.

### CODE REVIEW

File: src/Compilers/Core/Portable/InternalUtilities/SemaphoreSlimExtensions.cs [hunk 1]
```diff
@@ -35,7 +35,23 @@ public SemaphoreDisposer(SemaphoreSlim semaphore)
...
-                _semaphore.Release();
+                // Officially, Dispose() being called more than once is allowable, but in this case
+                // if that were to ever happen that means something is very, very wrong. Since it's an internal
+                // type, better to be strict.
+
+                // Nulling this out also means it's a bit easier to diagnose some async deadlocks; if you have an
+                // async deadlock where a SemaphoreSlim is held but you're unsure why, as long all the users of the
+                // SemaphoreSlim used the Disposable helpers, you can search memory and find the instance that
+                // is pointing to the SemaphoreSlim that hasn't nulled out this field yet; in that case you know
+                // that's holding the lock and can figure out who is holding that SemaphoreDisposer.
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/InternalUtilities/SemaphoreSlimExtensions.cs:46]: we're in a struct.  so this does only work as long as the NonCopyable part above really is maintained.  i can allow this :)

**Review by CyrusNajmabadi — APPROVED** [review]

**Review by ryzngard — APPROVED** [review]

**Review by 333fred — APPROVED** [review]

**Review by sharwell — APPROVED** [review]

File: src/Compilers/Core/Portable/InternalUtilities/SemaphoreSlimExtensions.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/Compilers/Core/Portable/InternalUtilities/SemaphoreSlimExtensions.cs:46]: @CyrusNajmabadi Indeed, the [NonCopyable] being violated does mean we could end up with "surprises", but at this point, this is the least of your concerns IMHO. :smile:

**Comment by 333fred** [comment]
@jasonmalinowski compiler code needs 2 sign offs from the team

**Comment by jasonmalinowski** [comment]
@333fred Oops, thanks. I was staring at this going "oh, I should get a second compiler review" and didn't realize I had already set auto-merge.

File: src/Compilers/Core/Portable/InternalUtilities/SemaphoreSlimExtensions.cs (hunk 1 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/InternalUtilities/SemaphoreSlimExtensions.cs:46]: > so this does only work as long as the NonCopyable part above really is maintained

The type is copied twice in this very file so `[NonCopyable]` is more like `[NonCopyableForSomeVersionsOfCopy]`

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/Core/Portable/InternalUtilities/SemaphoreSlimExtensions.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/InternalUtilities/SemaphoreSlimExtensions.cs:46]: > Instances of this type are copied twice in this very file

There are no copies of this type in this file (moves are not copies).

Comment (jaredpar) [reply src/Compilers/Core/Portable/InternalUtilities/SemaphoreSlimExtensions.cs:46]: There are two returns which are copies. You label them moves but that doesn't change that copies occur.

**Pull request merged**